engine/execution/ingestion/block_queue/queue.go

```diff
@@ -0,0 +1,382 @@
+// 1. the block has been validated by consensus algorithm
+// 2. the block's parent has been executed
+// 3. all the collections included in the block have been received
+type BlockQueue struct {
+	sync.Mutex
+	log zerolog.Logger
+
+	// if a block still exists in this map, it means either some of its collection is missing,
+	// or its parent block has not been executed.
+	// if a block's StartState is not nil, it means its parent block has been executed, and
+	// its parent block must have been removed from this map
+	// if a block's StartState is nil, it means its parent block has not been executed yet.
+	// and its parent must be found in the this map as well
+	blocks map[flow.Identifier]*entity.ExecutableBlock
+
+	// a collection could be included in multiple blocks,
+	// when a missing block is received, it might trigger multiple blocks to be executable, which
+	// can be looked up by the map
+	// when a block is executed, its collections should be removed from this map unless a collection
+	// is still referenced by other blocks, which will eventually be removed when those blocks are
+	// executed.
+	collections map[flow.Identifier]*collectionInfo
+
+	// blockIDsByHeight is used to find next executable block.
+	// when a block is executed, the next executable block must be a block with height = current block height + 1
+	// the following map allows us to find the next executable block by height and their parent block ID
+	blockIDsByHeight map[uint64]map[flow.Identifier]*entity.ExecutableBlock
+}
+
+type MissingCollection struct {
+	BlockID   flow.Identifier
+	Height    uint64
+	Guarantee *flow.CollectionGuarantee
+}
+
+func (m *MissingCollection) ID() flow.Identifier {
+	return m.Guarantee.ID()
+}
+
+// collectionInfo is an internal struct used to keep track of the state of a collection,
+// and the blocks that include the collection
+type collectionInfo struct {
+	Collection *entity.CompleteCollection
+	IncludedIn map[flow.Identifier]*entity.ExecutableBlock
+}
+
+func NewBlockQueue(logger zerolog.Logger) *BlockQueue {
+	log := logger.With().Str("module", "block_queue").Logger()
+
+	return &BlockQueue{
+		log:              log,
+		blocks:           make(map[flow.Identifier]*entity.ExecutableBlock),
+		collections:      make(map[flow.Identifier]*collectionInfo),
+		blockIDsByHeight: make(map[uint64]map[flow.Identifier]*entity.ExecutableBlock),
+	}
+}
+
+// OnBlock is called when a new block is received, the parentFinalState indicates
+// whether its parent block has been executed.
+// Caller must ensure:
+// 1. blocks are passsed in order, i.e. parent block is passed in before its child block
+// 2. if a block's parent is not executed, then the parent block must be passed in first
+// 3. if a block's parent is executed, then the parent's finalState must be passed in
+// It returns (nil, nil, nil) if this block is a duplication
+func (q *BlockQueue) OnBlock(block *flow.Block, parentFinalState *flow.StateCommitment) (
```

Comment:
The `On*` naming is a little confusing since we typically use that for non-blocking signal functions. Here they are blocking and return some data. How about renaming them to something like `HandleBlock`?